File: heroku-jvm/src/main/java/com/riseofcat/server/PingDecorator.java
package com.riseofcat.server;
import com.riseofcat.lib.TypeMap;
import com.riseofcat.share.ClientSay;
import com.riseofcat.share.ServerSay;

import org.jetbrains.annotations.Nullable;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class PingDecorator<C, S> extends SesServ<ClientSay<C>, ServerSay<S>> {
private final int pingIntervalMs;
private final SesServ<C, S> server;
private final Map<Ses, PingSes> map = new ConcurrentHashMap<>();
public PingDecorator(SesServ<C, S> server, int pingIntervalMs) {
	this.pingIntervalMs = pingIntervalMs;
	this.server = server;
}
public void start(Ses session) {
	PingSes s = new PingSes(session);
	map.put(session, s);
	server.start(s);
}
public void close(Ses session) {
	server.close(map.get(session));
	map.remove(session);
}
public void message(Ses session, ClientSay<C> say) {
	PingSes s = map.get(session);
	if(say.getPong() && s.lastPingTime != null) {
		long l = (System.currentTimeMillis() - s.lastPingTime + 1) / 2;
		s.latency = (int) l;
	}
	if(say.getPayload() != null) {
		server.message(s, say.getPayload());
	}
}

private class PingSes extends SesServ<C, S>.Ses {
	private final Ses sess;
	@Nullable private Long lastPingTime;
	@Nullable private Integer latency;
	private PingSes(Ses sess) {
		this.sess = sess;
		put(new Extra(this));
	}
	public int getId() {
		return sess.getId();
	}
	public void stop() {
		sess.stop();
	}
	public void send(S payload) {
		ServerSay<S> say = new ServerSay<S>(payload);
		say.setLatency(latency);
		if(lastPingTime == null || System.currentTimeMillis() > lastPingTime + pingIntervalMs) {
			say.setPing(true);
			lastPingTime = System.currentTimeMillis();
		}
		sess.send(say);
	}
	public TypeMap getTypeMap() {
		return sess.getTypeMap();
	}
}

public static class Extra implements TypeMap.Marker{
	private final PingDecorator.PingSes pingSes;
	public Extra(PingDecorator.PingSes pingSes) {
		this.pingSes = pingSes;
	}
	@Nullable public Integer getLatency() {
		return pingSes.latency;
	}
}

}

File: heroku-jvm/src/main/java/com/riseofcat/server/PingDecorator.kt
package com.riseofcat.server

import com.riseofcat.lib.*
import com.riseofcat.share.ping.*
import java.util.concurrent.ConcurrentHashMap

class PingDecorator<C,S>(private val server:SesServ<C,S>,private val pingInterval:Duration):SesServ<ClientSay<C>,ServerSay<S>>() {
  private val map = ConcurrentHashMap<Ses<ServerSay<S>>,PingSes>()
  override fun start(session:Ses<ServerSay<S>>) {
    val s = PingSes(session)
    map[session] = s
    server.start(s)
  }

  override fun close(session:Ses<ServerSay<S>>) {
    server.close(map[session]!!)
    map.remove(session)
  }

  override fun message(session:Ses<ServerSay<S>>,say:ClientSay<C>) {
    val s = map[session] ?: lib.log.fatalError("session not found")
    if(say.pong) {
      val pingDelay = Duration(1)+(lib.time-s.lastPing)/2
      s.pingDelay = pingDelay
      session send ServerSay(pingDelay = pingDelay, serverTime = lib.time)
    }
    if(say.payload!=null) server.message(s,say.payload!!)
  }

  inner class PingSes constructor(private val sess:Ses<ServerSay<S>>):Ses<S>() {
    var lastPing:TimeStamp = TimeStamp(0)
    var pingDelay:Duration? = null
    override val id:Int get() = sess.id
    override val typeMap:TypeMap get() = sess.typeMap

    init {
      put(Extra(this))
    }

    override fun stop() = sess.stop()

    override fun send(payload:S) {
      val ping =
        if(lib.time>lastPing+pingInterval) {
          lastPing = lib.time
          true
        } else false
      sess.send(ServerSay(payload,ping = ping))
    }
  }

  inner class Extra(private val pingSes:PingSes):TypeMap.Marker {
    val lastPingDelay:Duration? get() = pingSes.pingDelay//нужно будет для подсчёта средней pingDelay на этом сервере, чтобы фильтровать клиентов если они будут тормозить и переводить в другую комнату с лагающими
  }

}
